code/modules/projectiles/guns/gunhud.dm

```diff
@@ -0,0 +1,300 @@
+	var/obj/item/ammo_casing/energy/current_mode = the_gun.chambered
+	if(!current_mode)
+		return FALSE
+	return round(the_gun.cell.charge/current_mode.e_cost)
+
+/datum/component/ammo_hud/proc/update_hud()
+	SIGNAL_HANDLER
+	var/obj/item/gun/ballistic/pew = parent
+	hud.maptext = null
+	hud.icon_state = "[prefix]backing"
+	if(!pew.magazine)
+		hud.set_hud(backing_color, "[prefix]oe", "[prefix]te", "[prefix]he", "[prefix]no_mag")
+		return
+	if(!pew.get_ammo())
+		hud.set_hud(backing_color, "[prefix]oe", "[prefix]te", "[prefix]he", "[prefix]empty_flash")
+		return
+
+	var/indicator
+	var/rounds = num2text(get_accurate_ammo_count(pew))
+	var/oth_o
+	var/oth_t
+	var/oth_h
+
+	switch(length(rounds))
+		if(1)
+			oth_o = "[prefix]o[rounds[1]]"
+		if(2)
+			oth_o = "[prefix]o[rounds[2]]"
+			oth_t = "[prefix]t[rounds[1]]"
+		if(3)
+			oth_o = "[prefix]o[rounds[3]]"
+			oth_t = "[prefix]t[rounds[2]]"
+			oth_h = "[prefix]h[rounds[1]]"
+		else
+			oth_o = "[prefix]o9"
+			oth_t = "[prefix]t9"
+			oth_h = "[prefix]h9"
+	hud.set_hud(backing_color, oth_o, oth_t, oth_h, indicator)
+
+/datum/component/ammo_hud/laser/update_hud()
+	var/obj/item/gun/energy/pew = parent
+	hud.maptext = null
+	hud.icon_state = "[prefix]backing"
+	if(!pew.cell)
+		hud.set_hud(backing_color, "[prefix]oe", "[prefix]te", "[prefix]he", "[prefix]no_mag")
+		return
+	if(!get_accurate_laser_count(pew))
+		hud.set_hud(backing_color, "[prefix]oe", "[prefix]te", "[prefix]he", "[prefix]empty_flash")
+		return
+
+	var/indicator
+	var/rounds = num2text(get_accurate_laser_count(pew))
+	var/oth_o
+	var/oth_t
+	var/oth_h
+
+	switch(length(rounds))
+		if(1)
+			oth_o = "[prefix]o[rounds[1]]"
+		if(2)
+			oth_o = "[prefix]o[rounds[2]]"
+			oth_t = "[prefix]t[rounds[1]]"
+		if(3)
+			oth_o = "[prefix]o[rounds[3]]"
+			oth_t = "[prefix]t[rounds[2]]"
+			oth_h = "[prefix]h[rounds[1]]"
+		else
+			oth_o = "[prefix]o9"
+			oth_t = "[prefix]t9"
+			oth_h = "[prefix]h9"
+	hud.set_hud(backing_color, oth_o, oth_t, oth_h, indicator)
+
+/datum/component/ammo_hud/laser/cybersun
+	prefix = "cybersun_"
+
+/datum/component/ammo_hud/revolver
+	prefix = "revolver_"
+
+/datum/component/ammo_hud/revolver/turn_on()
+	var/obj/item/gun/ballistic/revolver/gun = parent
+	if(!HAS_TRAIT(gun.loc, TRAIT_GUNSLINGER))
+		return
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, PROC_REF(turn_off))
+	RegisterSignal(parent, list(COMSIG_UPDATE_AMMO_HUD, COMSIG_GUN_CHAMBER_PROCESSED), PROC_REF(update_hud))
+
+	hud.turn_on()
+	update_hud()
+
+
+/// Returns get_ammo() with the appropriate args passed to it - some guns like the revolver and bow are special cases
+/datum/component/ammo_hud/revolver/get_accurate_ammo_count(obj/item/gun/ballistic/revolver/the_gun)
+	if(istype(the_gun, /obj/item/gun/ballistic/revolver))
+		var/obj/item/gun/ballistic/revolver/the_revolver = the_gun
+		if(the_revolver.magazine)
+			return the_revolver.magazine.ammo_list()
+		else
+			return FALSE
```

Comment:
Probably can get away with killing this else branch and just allowing this proc to return null